src/flint/types/fmpz_mod_mpoly.pyx

```diff
@@ -0,0 +1,1192 @@
+            return res
+        return NotImplemented
+
+    def __pow__(self, other, modulus):
+        cdef fmpz_mod_mpoly res
+        if modulus is not None:
+            raise NotImplementedError
+        other = any_as_fmpz(other)
+        if other is NotImplemented:
+            return other
+        if other < 0:
+            raise ValueError("cannot raise fmpz_mod_mpoly to negative power")
+        res = create_fmpz_mod_mpoly(self.ctx)
+        if fmpz_mod_mpoly_pow_fmpz(res.val, (<fmpz_mod_mpoly>self).val, (<fmpz>other).val, res.ctx.val) == 0:
+            raise ValueError("unreasonably large polynomial")  # pragma: no cover
+        return res
+
+    def __divmod__(self, other):
+        cdef fmpz_mod_mpoly res, res2
+
+        if not self.ctx.is_prime():
+            raise DomainError("division with non-prime modulus is not supported")
+        elif typecheck(other, fmpz_mod_mpoly):
+            if not other:
+                raise ZeroDivisionError("fmpz_mod_mpoly division by zero")
+            elif (<fmpz_mod_mpoly>self).ctx is not (<fmpz_mod_mpoly>other).ctx:
+                raise IncompatibleContextError(f"{(<fmpz_mod_mpoly>self).ctx} is not {(<fmpz_mod_mpoly>other).ctx}")
+            res = create_fmpz_mod_mpoly(self.ctx)
+            res2 = create_fmpz_mod_mpoly(self.ctx)
+            fmpz_mod_mpoly_divrem(res.val, res2.val, (<fmpz_mod_mpoly>self).val, (<fmpz_mod_mpoly>other).val, res.ctx.val)
+            return (res, res2)
+        else:
+            other = any_as_fmpz(other)
+            if other is not NotImplemented:
+                other = fmpz_mod_mpoly(other, self.ctx)
+                if not other:
+                    raise ZeroDivisionError("fmpz_mod_mpoly division by zero")
+                res = create_fmpz_mod_mpoly(self.ctx)
+                res2 = create_fmpz_mod_mpoly(self.ctx)
+                fmpz_mod_mpoly_divrem(res.val, res2.val, (<fmpz_mod_mpoly>self).val, (<fmpz_mod_mpoly>other).val, res.ctx.val)
+                return (res, res2)
+        return NotImplemented
+
+    def __rdivmod__(self, other):
+        cdef fmpz_mod_mpoly res, res2
+        if not self.ctx.is_prime():
+            raise DomainError("division with non-prime modulus is not supported")
+        elif not self:
+            raise ZeroDivisionError("fmpz_mod_mpoly division by zero")
+        other = any_as_fmpz(other)
+        if other is not NotImplemented:
+            other = fmpz_mod_mpoly(other, self.ctx)
+            res = create_fmpz_mod_mpoly(self.ctx)
+            res2 = create_fmpz_mod_mpoly(self.ctx)
+            fmpz_mod_mpoly_divrem(res.val, res2.val, (<fmpz_mod_mpoly>other).val, (<fmpz_mod_mpoly>self).val, res.ctx.val)
+            return (res, res2)
+        return NotImplemented
+
+    def __floordiv__(self, other):
+        cdef fmpz_mod_mpoly res
+        if not self.ctx.is_prime():
+            raise DomainError("division with non-prime modulus is not supported")
```

Comment:
This applies to other cases as well. We need to check the type of `other` and return `NotImplemented` before checking anything else. It is important to return `NotImplemented` rather than raise a `TypeError` because that is how Python's cooperative dispatch works for binary operators. If we return `NotImplemented` here then it gives a chance for the other type to implement `__rfloordiv__`. Other future types might (in principle) want to define this operation with this type for whatever reason even if the modulus is not prime.